### DISCUSSION TITLE
More robust Authentication API

### DISCUSSION THREAD

**Discussion by Darhagonable** [expo/expo#27722] [Proposals and Ideas]
The current expo library for authentication is expo-auth-session but in my use cases it seems pretty lacking or at least not that clearly documented for how to make full solution.

Would be nice the the API was more akin to the likes of [oidc-react](https://github.com/bjerkio/oidc-react) or [react-oidc-context](https://github.com/authts/react-oidc-context)

### Set of requitements:
- Option to automatically renew the access token prior to its expiration
- Option to automatically login when launching app using refresh-token from last session
- Option to support local/offline-first by persisting login across app launches without needing to make network requests
- Option to be redirected back to previous route after signing in.

### Nice to haves:
- Support Expo Go (at least enough for testing purposes)
- Option to support multiple providers/tenants (sometimes one has one IDP for customers and another for employees)
- Possibly allow to open the login pages within a webview instead of navigating out of the app.

### Proposed API:
Have API similar to [oidc-react](https://github.com/bjerkio/oidc-react) and [react-oidc-context](https://github.com/authts/react-oidc-context) with a simple React Context Provider and then a useAuth() hook to get the `userData`, `accessToken`, `signIn` method, `signOut` method and more

### How to support multiple providers/tenants:
Provider:
```jsx
<AuthProvider
  configs={{
    auth0Employees: {
      domain,
      clientId,
      audience,
      ...
    },
    auth0Customers: {
      domain,
      clientId,
      audience,
      ...
    }
  }
>
  <App/>
</AuthProvider>
```
Sign In method:

```jsx
const { signIn } = useAuth();

return (
  < >
    <Button
      onPress={() => signIn("auth0Employees")}
      title="Sign in as employee"
    />
    <Button
      onPress={() => signIn("auth0Customers")}
      title="Sign in as customer"
    />
  </>
);
```

**Comment by idc77** [comment]
I agree, it's a long way from being useful.
I can't believe that people are fine with the way things are right now.
